src/windows/sid.rs

```diff
@@ -0,0 +1,168 @@
+        sddl::{ConvertSidToStringSidW, ConvertStringSidToSidW},
+        winerror::ERROR_INSUFFICIENT_BUFFER,
+    },
+    um::{
+        errhandlingapi::GetLastError,
+        securitybaseapi::{CopySid, EqualSid, GetLengthSid, IsValidSid},
+        winbase::{LocalFree, LookupAccountSidW},
+        winnt::{SidTypeUnknown, LPWSTR, PSID},
+    },
+};
+
+use crate::sys::utils::to_str;
+
+#[doc = include_str!("../../md_doc/sid.md")]
+#[derive(Clone, Debug)]
+pub struct Sid {
+    sid: Vec<u8>,
+}
+
+impl Sid {
+    /// Creates an `Sid` by making a copy of the given raw SID.
+    pub(crate) unsafe fn from_psid(psid: PSID) -> Option<Self> {
+        if psid.is_null() {
+            return None;
+        }
+
+        if IsValidSid(psid) == 0 {
+            return None;
+        }
+
+        let length = GetLengthSid(psid);
+
+        let mut sid = vec![0; length as usize];
+
+        if CopySid(length, sid.as_mut_ptr() as *mut _, psid) == 0 {
+            sysinfo_debug!("CopySid failed: {:?}", GetLastError());
+            return None;
+        }
+
+        Some(Self { sid })
+    }
+
+    /// Retrieves the account name of this SID.
+    pub(crate) fn account_name(&self) -> Option<String> {
+        unsafe {
+            let mut name_len = 0;
+            let mut domain_len = 0;
+            let mut name_use = SidTypeUnknown;
+
+            if LookupAccountSidW(
+                std::ptr::null_mut(),
+                self.sid.as_ptr() as *mut _,
+                std::ptr::null_mut(),
+                &mut name_len,
+                std::ptr::null_mut(),
+                &mut domain_len,
+                &mut name_use,
+            ) == 0
+            {
+                let error = GetLastError();
+                if error != ERROR_INSUFFICIENT_BUFFER {
+                    sysinfo_debug!("LookupAccountSidW failed: {:?}", error);
+                    return None;
+                }
+            }
+
+            let mut name = vec![0; name_len as usize];
+
+            // Reset length to 0 since we're still passing a NULL pointer
+            // for the domain.
+            domain_len = 0;
+
+            if LookupAccountSidW(
+                std::ptr::null_mut(),
+                self.sid.as_ptr() as *mut _,
+                name.as_mut_ptr(),
+                &mut name_len,
+                std::ptr::null_mut(),
+                &mut domain_len,
+                &mut name_use,
+            ) == 0
+            {
+                sysinfo_debug!("LookupAccountSidW failed: {:?}", GetLastError());
+                return None;
+            }
+
+            Some(to_str(name.as_mut_ptr()))
+        }
+    }
+}
+
+impl PartialEq for Sid {
+    fn eq(&self, other: &Self) -> bool {
+        unsafe { EqualSid(self.sid.as_ptr() as *mut _, other.sid.as_ptr() as *mut _) != 0 }
+    }
+}
+
+impl Eq for Sid {}
+
+impl PartialOrd for Sid {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.to_string().partial_cmp(&other.to_string())

```

Comment:
Let's avoid allocations in here by using [from_utf8_unchecked](https://doc.rust-lang.org/nightly/std/str/fn.from_utf8_unchecked.html).

Reply:
Just thought: what's the added value of converting into string before making the comparison btw?

Reply:
The problem I'm trying to work around here is that the SID structure is not _technically_ documented. I mean, there's [this](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-sid), but in the same place they also say:

> Applications should not modify a SID directly. To create and manipulate a security identifier, use the functions listed in the See Also section.

So the only somewhat reliable way to do this is by converting to a string, which can be sorted and hashed.

The implicit assumption here is that the string representation returned by [`ConvertSidToStringSidW`](https://learn.microsoft.com/en-us/windows/desktop/api/sddl/nf-sddl-convertsidtostringsidw) is "canonical". So that if I call this function on two SIDs that compare equal according to `EqualSid`, then their string representations will also compare equal.

I think it's a pretty safe assumption, since these strings have been around for basically forever.

But maybe I'm just being overly pedantic here. For as long as SIDs have existed there has only ever been one revision of the `SID` structure, and its format is documented. We could avoid all this pain by just comparing/hashing the `Vec`s directly. After all, that's what `EqualSid` does: compute the size of each SID and call `memcmp`. So the stability of the format of a SID _string_ is probably the same as the stability of the binary structure itself.

What do you think?

Reply:
If we remove the `char` wrapping, it's basically the same I think. In both cases, the `PartialOrd` implementation is just here so that we can have vectors of `SID`. But we can keep it as is and just use `from_utf8_unchecked` if you want to keep the `str` wrapping, I'm fine with it. Let's just avoid the unneeded allocation from `to_string()`.

Reply:
Okay, how about this: we can get rid of all the weird hacks and just assert that the SID has the revision we expect (1). If it changes (which I don't expect it will), we'll catch it.

Reply:
Sounds good to me. If it does become problematic, we can always go for another implementation later on in any case.